### PR TITLE
Allow a Task to be actioned it if EXTENDS an incomplete Task

### LINKED ISSUE #84: Allow a Task to be actioned it if EXTENDS an incomplete Task?
If a user wishes to create a chain of Tasks for a given Transformation that EXTEND one another in a sequence, currently only the first Task in the chain (which would *not* EXTEND an incomplete Task) would be actionable. This then requires that users action Tasks in the chain individually as they execute. This is painful at best.

Instead, we can remove the filter for Tasks that EXTEND only a complete Task, allowing any Task to be actioned regardless of this. The `claim_taskhub_task` method would then need to perform this filtering instead. 

This should overall give a better user experience, allowing users to fire and forget the whole chain at once.

### PR DESCRIPTION
**Pull request by hmacdope**
Fixes #84 
- modified the `action_tasks` code to allow actioning of tasks via an `EXTENDS` relationship even if not complete 
- Equivalent changes to the `claim` code to only allow claiming of extends tasks if the preceding node is complete.

### CODE REVIEW

**Comment by hmacdope** [comment]
@dotsdl I think this is ready for review. There is just one failing test which Is possibly on the GUFE side and is being triggered as we now have EXTENDS tasks.

**Comment by hmacdope** [comment]
Specifically GUFE is complaining about the lack of a `protocol_unit_result` method. It may be a GUFE versioning thing, but thought best to discuss.

**Review by dotsdl — CHANGES_REQUESTED** [review]
@hmacdope looking good, but one major note that needs addressing before we proceed. Should be able to drop consideration of EXTENDS in `action_tasks`.

File: alchemiscale/storage/statestore.py [hunk 1]
```diff
@@ -859,29 +859,47 @@ def action_tasks(
...
+                // now we need to check if the task extends another task
+                // call allows us to return the results to the outer scope
+                CALL {{
+                WITH th, an, task
+                    MATCH (task)-[:EXTENDS]->(extends_task:Task)
+                    WHERE NOT (th)-[:ACTIONS]->(extends_task) OR extends_task IS NULL
+                    CREATE (th)-[ar_e:ACTIONS {{weight: 1.0}}]->(extends_task)
+                    SET ar_e.task = extends_task._scoped_key
+                    return collect(extends_task) as extends 
+                }}
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply alchemiscale/storage/statestore.py:890]: I don't know why I had in my head that we wanted to `ACTION` any tasks that `EXTEND` the target task. My bad, I'll fix it up ASAP.

File: alchemiscale/storage/statestore.py [hunk 2]
```diff
@@ -859,29 +859,47 @@ def action_tasks(
...
-                    ScopedKey.from_str(task["_scoped_key"])
-                    if task is not None
-                    else None
-                )
+                tasks = tx.run(q)
+                # ugly nesting because of `collect`
+                for stream in tasks:
+                    for graph in stream:
+                        for record in graph:
+                            if record is not None:
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply alchemiscale/storage/statestore.py:901]: Yep makes sense.

File: alchemiscale/storage/statestore.py [hunk 3]
```diff
@@ -859,29 +859,47 @@ def action_tasks(
...
+                CALL {{
+                WITH th, an, task
+                    // create the connections 
+                    // set the ACTIONS relationship weight to default weight of 1.0
+                    CREATE (th)-[ar:ACTIONS {{weight: 1.0}}]->(task)
+                    // set the task property to the scoped key of the Task
+                    // this is a convenience for when we have to loop over relationships in Python
+                    SET ar.task = task._scoped_key
+                    RETURN collect(task) as main_task
+                }}
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply alchemiscale/storage/statestore.py]: `COLLECT` makes it available in the outer scope, but the whole thing should simplify when we remove any EXTENDS funky business.

**Comment by hmacdope** [comment]
Fixed @dotsdl sorry about that.

**Review by dotsdl — APPROVED** [review]
Looks great @hmacdope! Thank you!

Merging!

**Comment by dotsdl** [comment]
Fixing CI; will get it sorted then merge.

**Pull request merged**